common/rfb/CSecurityRSAAES.cxx

```diff
@@ -0,0 +1,444 @@
+  }
+  return NULL;
+}
+
+static void random_func(void* ctx, size_t length, uint8_t* dst)
+{
+  gnutls_rnd(GNUTLS_RND_RANDOM, dst, length);
+}
+
+void CSecurityRSAAES::writePublicKey()
+{
+  // generate client key
+  rsa_public_key_init(&clientPublicKey);
+  rsa_private_key_init(&clientKey);
+  int rsaKeySize = (clientKeyLength + 7) / 8;
+  // set key size to non-zero to allow clearing the keys when cleanup
+  clientPublicKey.size = rsaKeySize;
+  clientKey.size = rsaKeySize;
+  // set e = 65537
+  mpz_set_ui(clientPublicKey.e, 65537);
+  if (!rsa_generate_keypair(&clientPublicKey, &clientKey,
+    NULL, random_func, NULL, NULL, clientKeyLength, 0))
+    throw AuthFailureException("failed to generate key");
+  clientKeyN = new rdr::U8[rsaKeySize];
+  clientKeyE = new rdr::U8[rsaKeySize];
+  nettle_mpz_get_str_256(rsaKeySize, clientKeyN, clientPublicKey.n);
+  nettle_mpz_get_str_256(rsaKeySize, clientKeyE, clientPublicKey.e);
+  rawos->writeU32(clientKeyLength);
+  rawos->writeBytes(clientKeyN, rsaKeySize);
+  rawos->writeBytes(clientKeyE, rsaKeySize);
+  rawos->flush();
+}
+
+bool CSecurityRSAAES::readPublicKey()
+{
+  if (!rawis->hasData(4))
+    return false;
+  rawis->setRestorePoint();
+  serverKeyLength = rawis->readU32();
+  if (serverKeyLength < MinKeyLength)
+    throw AuthFailureException("server key is too short");
+  if (serverKeyLength > MaxKeyLength)
+    throw AuthFailureException("server key is too long");
+  size_t size = (serverKeyLength + 7) / 8;
+  if (!rawis->hasDataOrRestore(size * 2))
+    return false;
+  rawis->clearRestorePoint();
+  serverKeyE = new rdr::U8[size];
+  serverKeyN = new rdr::U8[size];
+  rawis->readBytes(serverKeyN, size);
+  rawis->readBytes(serverKeyE, size);
+  rsa_public_key_init(&serverKey);
+  nettle_mpz_set_str_256_u(serverKey.n, size, serverKeyN);
+  nettle_mpz_set_str_256_u(serverKey.e, size, serverKeyE);
+  if (!rsa_public_key_prepare(&serverKey))
+    throw AuthFailureException("server key is invalid");
+  return true;
+}
+
+void CSecurityRSAAES::verifyServer()
+{
+  rdr::U8 lenServerKey[4] = {
+    (rdr::U8)((serverKeyLength & 0xff000000) >> 24),
+    (rdr::U8)((serverKeyLength & 0xff0000) >> 16),
+    (rdr::U8)((serverKeyLength & 0xff00) >> 8),
+    (rdr::U8)(serverKeyLength & 0xff)
+  };
+  rdr::U8 f[8];
+  struct sha1_ctx ctx;
+  sha1_init(&ctx);
+  sha1_update(&ctx, 4, lenServerKey);
+  sha1_update(&ctx, serverKey.size, serverKeyN);
+  sha1_update(&ctx, serverKey.size, serverKeyE);
+  sha1_digest(&ctx, sizeof(f), f);
+  const char *title = "Server key fingerprint";
+  CharArray text;
+  text.format(
+    "The server has provided the following identifying information:\n"
+    "Fingerprint: %02x-%02x-%02x-%02x-%02x-%02x-%02x-%02x\n"
+    "Please verify that the information is correct and press \"Yes\". "
+    "Otherwise press \"No\"", f[0], f[1], f[2], f[3], f[4], f[5], f[6], f[7]);
+  if (!CSecurityTLS::msg->showMsgBox(UserMsgBox::M_YESNO, title, text.buf))
```

Comment:
This is an odd connection, so I think it is time to move this connection to `CSecurity` instead.